internal/providers/pluginfw/resources/sharing/resource_share.go

```diff
@@ -0,0 +1,396 @@
+	}
+
+	newState.SyncEffectiveFieldsDuringRead(existingState.ShareInfo)
+	for i := range newState.Objects {
+		newState.Objects[i].SyncEffectiveFieldsDuringRead(existingState.Objects[i])
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, newState)...)
+}
+
+func (r *ShareResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var state ShareInfoExtended
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	client, diags := r.Client.GetWorkspaceClient()
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var plan ShareInfoExtended
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var planGoSDK sharing.ShareInfo
+	resp.Diagnostics.Append(converters.TfSdkToGoSdkStruct(ctx, plan, &planGoSDK)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var getShareRequest sharing.GetShareRequest
+	getShareRequest.Name = state.Name.ValueString()
+	getShareRequest.IncludeSharedData = true
+
+	currentShareInfo, err := client.Shares.Get(ctx, getShareRequest)
+	if err != nil {
+		resp.Diagnostics.AddError("failed to get current share info", err.Error())
+		return
+	}
+
+	matchOrder(currentShareInfo.Objects, planGoSDK.Objects, func(obj sharing.SharedDataObject) string { return obj.Name })
+	suppressCDFEnabledDiff(currentShareInfo)
+
+	changes := diff(*currentShareInfo, planGoSDK)
+
+	// if owner has changed, update the share owner
+	if !plan.Owner.IsNull() {
+		updatedShareInfo, err := client.Shares.Update(ctx, sharing.UpdateShare{
+			Name:  state.Name.ValueString(),
+			Owner: plan.Owner.ValueString(),
+		})
+		if err == nil {
+			resp.Diagnostics.Append(converters.GoSdkToTfSdkStruct(ctx, updatedShareInfo, &state)...)
+			if resp.Diagnostics.HasError() {
+				return
+			}
+		} else {
+			resp.Diagnostics.AddError("failed to update share owner", err.Error())
+			return
+		}
+	}
+
+	if len(changes) > 0 {
+		// if there are any other changes, update the share with the changes
+		updatedShareInfo, err := client.Shares.Update(ctx, sharing.UpdateShare{
+			Name:    plan.Name.ValueString(),
+			Updates: changes,
+		})
+		if err == nil {
+			resp.Diagnostics.Append(converters.GoSdkToTfSdkStruct(ctx, updatedShareInfo, &state)...)
+			if resp.Diagnostics.HasError() {
+				return
+			}
+		} else {
+			rollbackShareInfo, rollbackErr := client.Shares.Update(ctx, sharing.UpdateShare{
+				Name:  currentShareInfo.Name,
+				Owner: currentShareInfo.Owner,
+			})
+			if rollbackErr == nil {
+				resp.Diagnostics.Append(converters.GoSdkToTfSdkStruct(ctx, rollbackShareInfo, &state)...)
+				if resp.Diagnostics.HasError() {
+					return
+				}
+			} else {
+				resp.Diagnostics.AddError("failed to roll back", common.OwnerRollbackError(err, rollbackErr, currentShareInfo.Owner, plan.Owner.ValueString()).Error())
+			}
+
+			resp.Diagnostics.AddError("failed to update share", err.Error())
```

Comment:
[nit] Can this come at the beginning of this `else` block? If so, we can make the code much more readable by reducing the overall level of indentation.

Ref: https://go.dev/wiki/CodeReviewComments#indent-error-flow

```go
resp.Diagnostics.AddError("failed to update share", err.Error())

rollbackShareInfo, rollbackErr := client.Shares.Update(ctx, sharing.UpdateShare{
	Name:  currentShareInfo.Name,
	Owner: currentShareInfo.Owner,
})
if rollbackErr != nil {
	resp.Diagnostics.AddError("failed to roll back", common.OwnerRollbackError(err, rollbackErr, currentShareInfo.Owner, plan.Owner.ValueString()).Error())
	return
}

resp.Diagnostics.Append(converters.GoSdkToTfSdkStruct(ctx, rollbackShareInfo, &state)...)
if resp.Diagnostics.HasError() {
	return
}
```